FuelLabs/fuels-ts
# feat: adding support for proxy contracts in `fuels deploy`

> Closes TS-632
 - Closes #3123 

# Release notes

In this release, we:

 - Added support for proxy contracts in `fuels deploy`

# Summary

This PR adds support for `proxy` contracts, following the:
 - [SRC-14: Owned Proxy Contract](https://github.com/FuelLabs/sway-standard-implementations/tree/a3eb2a6b093d349a84f379920df7eee9ad1ad5d9/src14/owned_proxy)

The CLI command `fuels deploy` mimics the behavior from `forc-deploy`:
- https://docs.fuel.network/docs/forc/plugins/forc_client/#proxy-contracts

> [!NOTE]
> For an overview, refer also to this [flowchart](https://github.com/FuelLabs/fuels-ts/issues/3123#issuecomment-2361392146).

# Trying it out

To test it out using this PR's tag:

```bash
pnpm create fuels@pr-3190 --pnpm proxy-test --no-install && \
  cd proxy-test && \
  pnpm install
```

Then, choose a contract and enable `proxy` on its closest `Forc.toml` file:

https://github.com/FuelLabs/fuels-ts/blob/0a809336e58a4e90aa18c5b5270d55ce92131536/packages/fuels/test/fixtures/workspace/contracts/upgradable/Forc.toml#L10-L11

> [!IMPORTANT]
> If all works well, when you start your app in dev mode via `fuels dev`, you'll see that although the contract is re-deployed after being modified, its ID _stays the same_ in the `contracts.json` file. This happens because that's the ID of the proxy, not the underlying target contract.

# Checklist

- [x] All **changes** are **covered** by **tests** (or not applicable)
- [x] All **changes** are **documented** (or not applicable)
- [x] I **reviewed** the **entire PR** myself (preferably, on GH UI)
- [x] I **described** all **Breaking Changes** (or there's none)
